File: tools/plan-url.mjs
#!/usr/bin/env node
/**
 * plan-url.mjs — turn an acceptance guide into a share link, or a share link back into a guide.
 *
 * Part of the Acceptance Review Toolkit. Node standard library only; no dependencies, no network.
 * The link grammar this implements is the contract in FORMAT.md §8 — the web viewer decodes the
 * exact same payload, so a link produced here opens the guide anywhere the viewer is hosted.
 *
 *   encode:  node tools/plan-url.mjs docs/acceptance/042-foo-acceptance.md
 *   decode:  node tools/plan-url.mjs --decode "https://…/acceptance#plan=…"
 */

import { readFileSync, writeFileSync } from "node:fs";
import { basename } from "node:path";
import { deflateRawSync, inflateRawSync } from "node:zlib";

const DEFAULT_BASE = process.env.ACCEPTANCE_VIEWER_URL || "https://prototype.bejasc.dev/acceptance";
const WARN_LEN = 8000; // links longer than this get chewed up by some chat clients / proxies

const USAGE = `plan-url.mjs — acceptance guide ⇄ share link (FORMAT.md §8)

  Encode:  node tools/plan-url.mjs <guide.md|-> [options]
  Decode:  node tools/plan-url.mjs --decode <url|payload> [--out <file.md>]

Options
  --base <url>     Viewer base URL          (default: ${DEFAULT_BASE})
                   or set ACCEPTANCE_VIEWER_URL
  --name <file>    Filename to carry in the link (default: the input's basename)
  --no-name        Omit the name parameter
  --query          Use "?plan=" instead of the default "#plan="
  --raw            Skip DEFLATE; encode the markdown as plain base64url
  --markdown       Print a ready-to-paste markdown link instead of a bare URL
  --json           Print { url, payload, title, items, bytes } as JSON
  --force          Encode even if the file doesn't look like a valid guide
  --quiet          Suppress the size summary on stderr
  --out <file>     (decode) Write the markdown to a file instead of stdout
  -h, --help       This message

The fragment form (#plan=) is the default on purpose: a fragment is never sent to the server, so
the guide never leaves the reader's browser and never lands in a web-server log.
`;

/* ---------- base64url ---------- */
const toB64Url = (buf) =>
  buf.toString("base64").replace(/\+/g, "-").replace(/\//g, "_").replace(/=+$/, "");
const fromB64Url = (s) =>
  Buffer.from(String(s).replace(/\s+/g, "").replace(/-/g, "+").replace(/_/g, "/"), "base64");

/* ---------- guide validation (FORMAT.md §2) ---------- */
function inspect(md) {
  const lines = md.split(/\r?\n/);
  const titleLine = lines.find((l) => /^#\s+\S/.test(l));
  const title = titleLine ? titleLine.replace(/^#\s+/, "").trim() : null;

  const cells = (line) => {
    const t = line.trim();
    if (!(t.startsWith("|") && t.endsWith("|") && t.length > 2)) return null;
    return t.slice(1, -1).split(/(?<!\\)\|/).map((c) => c.trim());
  };
  const isSep = (c) => c && c.every((x) => /^:?-{2,}:?$/.test(x));

  let items = 0, found = false;
  for (let i = 0; i < lines.length; i++) {
    const c = cells(lines[i]);
    if (!(c && c.length >= 2 && /^id$/i.test(c[0]) && /verdict/i.test(c[c.length - 1]))) continue;
    if (!isSep(cells(lines[i + 1]))) continue;
    found = true;
    for (let j = i + 2; j < lines.length; j++) {
      const r = cells(lines[j]);
      if (!r || isSep(r)) break;
      items++;
    }
    break;
  }
  const reason = !title ? 'no "# " title heading'
    : !found ? "no verdict table (a header row starting with ID and ending with Verdict)"
    : items === 0 ? "the verdict table has no item rows"
    : null;
  return { ok: !reason, reason, title, items };
}

/* ---------- payload ---------- */
function encodePayload(md, { raw = false } = {}) {
  const plain = Buffer.from(md, "utf8");
  if (raw) return toB64Url(plain);
  const packed = deflateRawSync(plain, { level: 9 });
  // A pathological guide could deflate larger than it started; keep whichever is shorter.
  return toB64Url(packed.length < plain.length ? packed : plain);
}

function decodePayload(payload) {
  const bytes = fromB64Url(payload);
  const tries = [];
  try { tries.push(inflateRawSync(bytes).toString("utf8")); } catch { /* not deflated */ }
  try { tries.push(new TextDecoder("utf-8", { fatal: true }).decode(bytes)); } catch { /* not utf-8 */ }
  for (const md of tries) if (inspect(md).ok) return md;
  return tries[0] ?? null;
}

function extractPayload(input) {
  const s = String(input).trim();
  const m = s.match(/[#?&]plan=([A-Za-z0-9\-_+/=%]+)/);
  if (m) return decodeURIComponent(m[1]);
  return s; // assume a bare payload
}

/* ---------- args ---------- */
function parseArgs(argv) {
  const o = { base: DEFAULT_BASE, positional: [] };
  for (let i = 0; i < argv.length; i++) {
    const a = argv[i];
    const val = () => {
      const v = argv[++i];
      if (v === undefined) { console.error(`plan-url: ${a} needs a value`); process.exit(2); }
      return v;
    };
    switch (a) {
      case "-h": case "--help": o.help = true; break;
      case "-d": case "--decode": o.decode = true; break;
      case "--base": o.base = val(); break;
      case "--name": o.name = val(); break;
      case "--no-name": o.noName = true; break;
      case "--out": o.out = val(); break;
      case "--query": o.query = true; break;
      case "--raw": o.raw = true; break;
      case "--markdown": o.markdown = true; break;
      case "--json": o.json = true; break;
      case "--force": o.force = true; break;
      case "--quiet": o.quiet = true; break;
      default:
        if (a.startsWith("-") && a !== "-") { console.error(`plan-url: unknown option ${a}`); process.exit(2); }
        o.positional.push(a);
    }
  }
  return o;
}

const readInput = (p) => readFileSync(p === "-" ? 0 : p, "utf8");

/* ---------- main ---------- */
const opts = parseArgs(process.argv.slice(2));

if (opts.help || (!opts.positional.length && process.stdin.isTTY)) {
  process.stdout.write(USAGE);
  process.exit(opts.help ? 0 : 2);
}

const input = opts.positional[0] ?? "-";

if (opts.decode) {
  const source = input === "-" ? readInput("-") : input;
  const md = decodePayload(extractPayload(source));
  if (md == null) { console.error("plan-url: could not decode that link — the payload isn't valid base64url."); process.exit(1); }
  const info = inspect(md);
  if (!info.ok && !opts.quiet) console.error(`plan-url: warning — decoded text isn't a valid guide (${info.reason}).`);
  if (opts.out) { writeFileSync(opts.out, md, "utf8"); if (!opts.quiet) console.error(`plan-url: wrote ${opts.out} (${info.items} items)`); }
  else process.stdout.write(md.endsWith("\n") ? md : md + "\n");
  process.exit(0);
}

const md = readInput(input);
const info = inspect(md);
if (!info.ok && !opts.force) {
  console.error(`plan-url: ${input} doesn't look like an acceptance guide — ${info.reason}.`);
  console.error("plan-url: fix it against FORMAT.md, or pass --force to encode anyway.");
  process.exit(2);
}

const payload = encodePayload(md, { raw: opts.raw });
const name = opts.noName ? null : (opts.name ?? (input === "-" ? null : basename(input)));
const base = opts.base.replace(/[#?].*$/, "");
const sep = opts.query ? "?" : "#";
const url = `${base}${sep}plan=${payload}` + (name ? `&name=${encodeURIComponent(name)}` : "");

if (opts.json) {
  process.stdout.write(JSON.stringify(
    { url, payload, title: info.title, items: info.items, bytes: Buffer.byteLength(md, "utf8"), length: url.length },
    null, 2) + "\n");
} else if (opts.markdown) {
  process.stdout.write(`[${(info.title || "Acceptance guide").replace(/[[\]]/g, "")} — open in the viewer](${url})\n`);
} else {
  process.stdout.write(url + "\n");
}

if (!opts.quiet) {
  const src = Buffer.byteLength(md, "utf8");
  const pct = Math.round((1 - url.length / src) * 100);
  console.error(`plan-url: ${info.items} items, ${src} B markdown → ${url.length} char link${pct > 0 ? ` (${pct}% shorter)` : ""}`);
  if (url.length > WARN_LEN)
    console.error(`plan-url: warning — ${url.length} chars is long for a link; some chat clients truncate past ~${WARN_LEN}. Send the file instead, or trim the guide.`);
}

File: tools/plan_url.py
#!/usr/bin/env python3
"""plan_url.py — turn an acceptance guide into a share link, or a share link back into a guide.

The Python twin of tools/plan-url.mjs, for environments without Node. Standard library only;
no dependencies, no network. Both implement the same contract — FORMAT.md §8 — so a link from
either one opens in the web viewer and decodes with either tool. (The two DEFLATE
implementations tune differently, so the payloads aren't byte-identical; the markdown is.)

  encode:  python tools/plan_url.py docs/acceptance/042-foo-acceptance.md
  decode:  python tools/plan_url.py --decode "https://.../acceptance#plan=..."
"""

import argparse
import base64
import json
import os
import re
import sys
import urllib.parse
import zlib

DEFAULT_BASE = os.environ.get("ACCEPTANCE_VIEWER_URL", "https://prototype.bejasc.dev/acceptance")
WARN_LEN = 8000  # links longer than this get chewed up by some chat clients / proxies


# ---------- base64url ----------
def to_b64url(raw: bytes) -> str:
    return base64.urlsafe_b64encode(raw).decode("ascii").rstrip("=")


def from_b64url(s: str) -> bytes:
    s = re.sub(r"\s+", "", s).replace("-", "+").replace("_", "/")
    return base64.b64decode(s + "=" * (-len(s) % 4))


# ---------- guide validation (FORMAT.md §2) ----------
CELL_SPLIT = re.compile(r"(?<!\\)\|")


def _cells(line: str):
    t = line.strip()
    if not (t.startswith("|") and t.endswith("|") and len(t) > 2):
        return None
    return [c.strip() for c in CELL_SPLIT.split(t[1:-1])]


def _is_sep(cells) -> bool:
    return bool(cells) and all(re.fullmatch(r":?-{2,}:?", c) for c in cells)


def inspect(md: str) -> dict:
    lines = md.splitlines()
    title = next((l[1:].strip() for l in lines if re.match(r"^#\s+\S", l)), None)

    found, items = False, 0
    for i, line in enumerate(lines):
        c = _cells(line)
        if not (c and len(c) >= 2 and re.fullmatch(r"id", c[0], re.I) and re.search(r"verdict", c[-1], re.I)):
            continue
        if i + 1 >= len(lines) or not _is_sep(_cells(lines[i + 1])):
            continue
        found = True
        for row in lines[i + 2:]:
            r = _cells(row)
            if not r or _is_sep(r):
                break
            items += 1
        break

    if not title:
        reason = 'no "# " title heading'
    elif not found:
        reason = "no verdict table (a header row starting with ID and ending with Verdict)"
    elif items == 0:
        reason = "the verdict table has no item rows"
    else:
        reason = None
    return {"ok": reason is None, "reason": reason, "title": title, "items": items}


# ---------- payload ----------
def encode_payload(md: str, raw: bool = False) -> str:
    plain = md.encode("utf-8")
    if raw:
        return to_b64url(plain)
    co = zlib.compressobj(9, zlib.DEFLATED, -zlib.MAX_WBITS)  # negative wbits → raw DEFLATE
    packed = co.compress(plain) + co.flush()
    # A pathological guide could deflate larger than it started; keep whichever is shorter.
    return to_b64url(packed if len(packed) < len(plain) else plain)


def decode_payload(payload: str):
    data = from_b64url(payload)
    tries = []
    try:
        tries.append(zlib.decompress(data, -zlib.MAX_WBITS).decode("utf-8"))
    except Exception:
        pass
    try:
        tries.append(data.decode("utf-8"))
    except Exception:
        pass
    for md in tries:
        if inspect(md)["ok"]:
            return md
    return tries[0] if tries else None


def extract_payload(text: str) -> str:
    s = text.strip()
    m = re.search(r"[#?&]plan=([A-Za-z0-9\-_+/=%]+)", s)
    return urllib.parse.unquote(m.group(1)) if m else s


# ---------- I/O (byte-faithful: never translate CRLF, in either direction) ----------
def read_text(path: str) -> str:
    if path == "-":
        return sys.stdin.buffer.read().decode("utf-8")
    with open(path, "rb") as fh:
        return fh.read().decode("utf-8")


def write_stdout(text: str) -> None:
    sys.stdout.buffer.write(text.encode("utf-8"))
    sys.stdout.buffer.flush()


# ---------- main ----------
def main() -> int:
    p = argparse.ArgumentParser(
        prog="plan_url.py",
        description="Acceptance guide <-> share link (FORMAT.md section 8).",
        epilog="The fragment form (#plan=) is the default on purpose: a fragment is never sent to "
               "the server, so the guide never leaves the reader's browser.",
    )
    p.add_argument("input", nargs="?", default="-", help="guide .md to encode, or the link to decode ('-' = stdin)")
    p.add_argument("-d", "--decode", action="store_true", help="decode a link/payload back to markdown")
    p.add_argument("--base", default=DEFAULT_BASE, help=f"viewer base URL (default: {DEFAULT_BASE})")
    p.add_argument("--name", help="filename to carry in the link (default: the input's basename)")
    p.add_argument("--no-name", action="store_true", help="omit the name parameter")
    p.add_argument("--out", help="(decode) write markdown to a file instead of stdout")
    p.add_argument("--query", action="store_true", help='use "?plan=" instead of the default "#plan="')
    p.add_argument("--raw", action="store_true", help="skip DEFLATE; encode as plain base64url")
    p.add_argument("--markdown", action="store_true", help="print a ready-to-paste markdown link")
    p.add_argument("--json", action="store_true", help="print url/payload/title/items as JSON")
    p.add_argument("--force", action="store_true", help="encode even if the file isn't a valid guide")
    p.add_argument("--quiet", action="store_true", help="suppress the size summary on stderr")
    a = p.parse_args()

    if a.decode:
        source = read_text("-") if a.input == "-" else a.input
        md = decode_payload(extract_payload(source))
        if md is None:
            print("plan_url: could not decode that link — the payload isn't valid base64url.", file=sys.stderr)
            return 1
        info = inspect(md)
        if not info["ok"] and not a.quiet:
            print(f"plan_url: warning — decoded text isn't a valid guide ({info['reason']}).", file=sys.stderr)
        if a.out:
            with open(a.out, "wb") as fh:
                fh.write(md.encode("utf-8"))
            if not a.quiet:
                print(f"plan_url: wrote {a.out} ({info['items']} items)", file=sys.stderr)
        else:
            write_stdout(md if md.endswith("\n") else md + "\n")
        return 0

    md = read_text(a.input)
    info = inspect(md)
    if not info["ok"] and not a.force:
        print(f"plan_url: {a.input} doesn't look like an acceptance guide — {info['reason']}.", file=sys.stderr)
        print("plan_url: fix it against FORMAT.md, or pass --force to encode anyway.", file=sys.stderr)
        return 2

    payload = encode_payload(md, raw=a.raw)
    name = None if a.no_name else (a.name or (None if a.input == "-" else os.path.basename(a.input)))
    base = re.sub(r"[#?].*$", "", a.base)
    url = f"{base}{'?' if a.query else '#'}plan={payload}"
    if name:
        url += "&name=" + urllib.parse.quote(name, safe="")

    if a.json:
        print(json.dumps({"url": url, "payload": payload, "title": info["title"],
                          "items": info["items"], "bytes": len(md.encode('utf-8')),
                          "length": len(url)}, indent=2))
    elif a.markdown:
        label = re.sub(r"[\[\]]", "", info["title"] or "Acceptance guide")
        print(f"[{label} — open in the viewer]({url})")
    else:
        print(url)

    if not a.quiet:
        src = len(md.encode("utf-8"))
        pct = round((1 - len(url) / src) * 100)
        extra = f" ({pct}% shorter)" if pct > 0 else ""
        print(f"plan_url: {info['items']} items, {src} B markdown → {len(url)} char link{extra}", file=sys.stderr)
        if len(url) > WARN_LEN:
            print(f"plan_url: warning — {len(url)} chars is long for a link; some chat clients truncate "
                  f"past ~{WARN_LEN}. Send the file instead, or trim the guide.", file=sys.stderr)
    return 0


if __name__ == "__main__":
    sys.exit(main())
